docs/design/jet/025-jet-job-member-pruning.md

```diff
@@ -0,0 +1,205 @@
+```
+
+![Scan + Scan -> HashJoin](https://svgshare.com/i/tTL.svg)
+
+On the picture above the basic units are processors, since the actual graph looks like `Scan -> Transform -> Aggregate`.
+So, here we can eliminate whole Member 2, since according to partition key condition only Member 1 contains required data
+and final destination point `Aggregate`. Moreover, it is a good example for further processor pruning optimization :
+
+#### Solution design details
+
+To support this behaviour, we need to extend the API. One option is add a method to PMS:
+```java
+interface ProcessorMetaSupplier extends IdentifiedDataSerializable {
+    // ...
+    default boolean doesWorkWithoutInput() {
+        return true;
+    }
+    // ...
+}
+```
+The default return value is `true` for backwards compatibility. This means that all current and new processors that
+do no work without an input will have to override this method.
+
+We propose to change the contract of `ProcessorMetaSupplier#get` return function so that it will be
+allowed to return `null` for addresses for which it does not want to deploy processors. But this will not be enough
+to completely eliminate a member from the execution.
+
+Next, we need to determine exact scope of the change : there are two main approaches were considered, 
+to shorten their meaning we mark them as 'heavyweight' and 'lightweight'. The difference between these two approaches 
+is an exact stage when we understand that some member may be pruned from job execution: 'heavyweight' approach tries 
+to be more generic and define it for any accept DAG, when 'lightweight' approach focus on hinting `ExecutionPlanBuilder`
+with additional meta-information from DAG and JobConfig constructed by SQL optimizer.
+
+##### Heavyweight approach
+
+The core idea here is that  `ExecutionPlanBuilder` would have detailed analysis of received DAG before execution plan
+creation and then, possibly, trying to optimize DAG and only then apply member pruning based. The requirements
+to prune member are :
+- each vertex in the DAG may work without an input;
+- DAG does not contain `distributed-broadcast` edges;
+
+As a pros, using this approach generify member pruning usage for both SQL, Pipeline API; allow usage for any connectors.
+As a cons, this approach brings high complexity in problem analysis, long development and big list of corner cases, which
+are discussed in chapter below.
+
+##### Lightweight approach
+
+Unlike a way described above, we can think that member pruning would be beneficial mostly for small jobs.
+The overwhelming majority of such jobs are generated by SQL, and our team effort is to replace Predicate API,
+where `PartitionPredicate` is available for partitioned data querying.
+For that, we can use SQL optimization phase also for determining partitions and for all relations: we extract information
+about required partitions to run to JobConfig, and then mark DAG as 'ableToPruneMembers'. This mark is an indicator
```

Comment:
`ableToPruneMembers` flags seems redundant. We can always do that due to variability of cluster topology.

Reply:
Done.